include/bluetooth/gatt.h

```diff
@@ -1648,6 +1648,15 @@ enum {
 	 */
 	BT_GATT_SUBSCRIBE_FLAG_WRITE_PENDING,
 
+	/** @brief Required security check for notification/indication
+	 *
+	 *  If set, indicates that application explicitly provided minimum
+	 *  required security level in min_security parameter. If this flag is
+	 *  not set minimum required security level is implicitly set to
+	 *  security level from the time of subscription.
+	 */
+	BT_GATT_SUBSCRIBE_FLAG_MIN_SECURITY,
```

Comment:
I can't really figure out what the best course of action is here.

1) With this flag, we will possibly break some applications, as we will automatically set the security to the current security when subscribing, and possibly thus discard expected notifications in existing applications. Especially some notifications like "Service Changed", which is not expected to be encrypted, may have a minimum security set to L2 or above, and thus applications would not get those on an unencrypted link. 
2) Without this flag, the `min_security` field will more likely than not be set to 0 (L0), and in case of dynamically allocated parameters, will likely be uninitialized in existing applications. 

Either way will work, but we just need to ensure that this is documented properly in a way that makes this visible. 

I am not sure what is needed to add an API change that will potentially break existing applications. Will need input from someone like @carlescufi or @jhedberg

Reply:
yes, I have same doubts on how this should be enabled (initial version was opt-in only)

Reply:
If we go with 1., then we need to follow [this procedure](https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html#introducing-incompatible-changes). If we went with 2., I think a note in the release notes would likely suffice.

Reply:
My personal preference would be to do 2., as that requires the least amount of effort to implement, and is much less disruptive, and will only affect applications that use dynamic allocation (without initializing the struct to 0), assuming that we only upgrade the security level when the value is 0 to L1 (instead of conn->sec).

Reply:
thinking about this a bit more and TBH I don't think that falling back to current level is something that would break existing application even if notification on that specific characteristic is allowed on lower security level.

rationale:
 - this is the case only for bonded devices and thus device should be authenticated anyway (otherwise you don't know if bond exists)
 - if devices were bonded but not LTK is stored (ie only CSRK or IRK), which would cause no encryption on reconnection than this is not the case since it is not security mode 1

Reply:
> thinking about this a bit more and TBH I don't think that falling back to current level is something that would break existing application even if notification on that specific characteristic is allowed on lower security level.
> 
> rationale:
> 
>     * this is the case only for bonded devices and thus device should be authenticated anyway (otherwise you don't know if bond exists)
> 
>     * if devices were bonded but not LTK is stored (ie only CSRK or IRK), which would cause no encryption on reconnection than this is not the case since it is not security mode 1

But if the implementation uses the `connected` callback to send notifications on reconnection with bonded device, wouldn't the application be able to send the notification before the link has been encrypted? 

e.g. current flow on a server

1. Bonded client has connected
2. Server sends notification if a value has changed, but before encrypting the link as the characteristic value does not require encryption
3. Client receives notification
4. Link is encrypted
5. Server sends notifications for characteristic values that need encryption

But with this change it becomes

1. Bonded client has connected
2. Server sends notification if a value has changed, but before encrypting the link as the characteristic value does not require encryption
3. Client discards notification since security is lower than what it saved
4. Link is encrypted
5. Server sends notifications for characteristic values that need encryption

Since we cannot control the remote server behavior, storing a security level higher than what the server expects, will make us discard notifications we otherwise would have received.

Reply:
the thing is that until you enable encryption you don't know if peer device is bonded

Reply:
anyway, I'm fine with whatever consensus would be, either 1) 2) or any other solution (Kconfig option?)

Reply:
> the thing is that until you enable encryption you don't know if peer device is bonded

`bt_addr_le_is_bonded` doesn't need encryption to work AFAIK

Reply:
I guess what @sjanc is getting at is that the only way to know if both sides (not just us) has the same LTK is to try to encrypt. `bt_addr_le_is_bonded()` likely only checks the local side.

Reply:
> I guess what @sjanc is getting at is that the only way to know if both sides (not just us) have the same LTK is to try to encrypt. `bt_addr_le_is_bonded()` likely only checks the local side.

True, but if the service characteristic does not require encryption, wouldn't you still send the notification on connection to a bonded device, even before encryption?

Reply:
I'm for option 2). As noted by @Thalley, using the current security will result in discarding notifications received on connect. And if we choose to infer `min_security`, we would have to store it persistently, right?